istioctl/cmd/kubeinject.go

```diff
@@ -43,6 +59,130 @@ const (
 	valuesConfigMapKey = "values"
 )
 
+type ExternalInjector struct {
+	client       kube.ExtendedClient
+	clientConfig *admissionregistration.WebhookClientConfig
+}
+
+func (e ExternalInjector) Inject(pod *corev1.Pod) ([]byte, error) {
+	cc := e.clientConfig
+	if cc == nil {
+		return nil, fmt.Errorf("external injector is not available")
+	}
+	var address string
+	if cc.URL != nil {
+		address = *cc.URL
+	}
+	if cc.Service != nil {
+		svc, err := e.client.CoreV1().Services(cc.Service.Namespace).Get(context.Background(), cc.Service.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		namespace, selector, err := polymorphichelpers.SelectorsForObject(svc)
+		if err != nil {
+			return nil, fmt.Errorf("cannot attach to %T: %v", svc, err)
+		}
+		sortBy := func(pods []*corev1.Pod) sort.Interface { return sort.Reverse(podutils.ActivePods(pods)) }
+		pod, _, err := polymorphichelpers.GetFirstPod(e.client.CoreV1(), namespace, selector.String(), timeout, sortBy)
+		if err != nil {
+			return nil, err
+		}
+		podPort := 15017
+		for _, v := range svc.Spec.Ports {
+			if v.Name == webhookPortname {
+				podPort = v.TargetPort.IntValue()
+			}
+		}
+		f, err := e.client.NewPortForwarder(pod.Name, pod.Namespace, "", 0, podPort)
+		if err != nil {
+			return nil, err
+		}
+		if err := f.Start(); err != nil {
+			return nil, err
+		}
+		address = fmt.Sprintf("https://%s%s", f.Address(), *cc.Service.Path)
+		defer func() {
+			f.Close()
+			f.WaitForStop()
+		}()
+	}
+	secret, err := e.client.CoreV1().Secrets(istioNamespace).Get(context.TODO(), CASecret, metav1.GetOptions{})
```

Comment:
Do we need to read secrets?  I thought the webhook could do all of its work with just the `clientConfig.caBundle` from the webhook configuration?  (I could be wrong, I haven't tried with external control plane...)

Reply:
great idea! Let me change it to use the one from webhookconfig directly.